prov/util/src/util_attr.c

```diff
@@ -56,11 +71,79 @@ static int fi_valid_addr_format(uint32_t prov_format, uint32_t user_format)
 	}
 }
 
+char *ofi_strdup_less_prefix(char *name, char *prefix)
+{
+	return strdup(name + strlen(prefix) + 1);
+}
+
+char *ofi_strdup_add_prefix(char *name, char *prefix)
+{
+	char *prefix_name;
+	size_t len;
+	int ret;
+
+	len = strlen(prefix) + strlen(name) + 2;
+	if (!(prefix_name = malloc(len)))
+		return NULL;
+	ret = snprintf(prefix_name, len, "%s_%s", prefix, name);
+	if (ret != len - 1)
+		goto err;
+	return prefix_name;
+err:
+	free(prefix_name);
+	return NULL;
+}
+
+int ofi_layered_prov_getinfo(uint32_t version, const char *node, const char *service,
+			uint64_t flags, const struct fi_provider *prov,
+			const struct fi_info *prov_info, struct fi_info *hints,
+			ofi_alter_layer_info_t alter_layer_info,
+			ofi_alter_base_info_t alter_base_info,
+			int get_base_info, struct fi_info **info)
+{
+	struct fi_info *base_hints = NULL, *base_info;
+	int ret;
+
+	ret = fi_check_info(prov, prov_info, hints, FI_MATCH_PREFIX);
+	if (ret)
+		return ret;
+
+	if (hints) {
+		if (!(base_hints = alter_layer_info(hints)))
+			return -FI_ENOMEM;
+	}
+
+	ret = fi_getinfo(version, node, service, flags, base_hints, &base_info);
+	if (ret)
+		goto out;
+
+	if (get_base_info) {
+		*info = base_info;
+	} else {
+		if (!(*info = alter_base_info(base_info))) {
+			ret = -FI_ENOMEM;
+			goto out;
+		}
+		fi_freeinfo(base_info);
+	}
+out:
+	fi_freeinfo(base_hints);
+	return ret;
+}
+
+int fi_check_name(char *user_name, char *prov_name, enum fi_match_type type)
+{
+	return (type == FI_MATCH_PREFIX) ?
+		strncasecmp(user_name, prov_name, strlen(prov_name)) :
+		strcasecmp(prov_name, user_name);
+}
```

Comment:
Nearly all of the fi_* functions return 0 on success.  I would negate the strcmp return values here to match that, rather than passing through the values.  Also declare as static.